### ISSUE TITLE
Time location not constrained in dataset yamls

### ISSUE THREAD

**Issue by JackLidge** [opendatacube/datacube-core#985]
`Datacube-core` version: 1.7
`OS`: Ubuntu 18.04

Linked to a resolved issue in opendatacube/datacube-ows#301. I was updating to a more recent version of the OWS when I discovered a lot of products were no longer being picked up by `update_ranges.py` in that repository.

After diagnosis, this turned out to be due to time parameters such as `from_dt` and `to_dt` in the indexed yaml file being located in the `grid_spatial` subsection instead of the `extents` subsection in the yaml. These products worked perfectly fine in the datacube and in our scripts we've written to manipulate data, but failed the update_ranges.py update.

It was then suggested I open an issue here to flag up that the datacube doesn't perform YAML schema validation for space and time to prevent such errors during indexing.

**Comment by Kirill888** [comment]
The problem with dataset yamls is that they are not constrained at all BY DESIGN, not all dataset documents describe geo referenced raster data. And the content of the dataset document is "relocatable" via metadata mechanism. So it's not feasible to have a simple schema that validates dataset document without re-designing indexing part and adding actual constraints.

Having said that `datacube dataset add` can and should perform more checks than it currently does, we have an issue here #955.

> These products worked perfectly fine in the datacube and in our scripts we've written to manipulate data, but failed the update_ranges.py update.

I find this hard to believe, I agree that datasets would index without warnings, but there would be no way to query them by time dimension if your time data was stored in the wrong location. We have not changed anything in 1.8 in that regard. Maybe you had some time information in the right spot, but not exactly the right format that ows expects? I'm pretty sure that OWS makes more assumptions about metadata structure than what is required by datacube itself. It is a constant source of pain to be frank.

Can you please provide examples of your dataset and product definition.

**Comment by Kirill888** [comment]
It looks like ows assumes that time information is recorded in `from_dt/to_dt` variables, but datacube EO metadata allows also for a single `center_dt` as well. So even if we had (and we should) had index time dataset validation logic implemented it would not have prevented that particular failure mode.

https://github.com/opendatacube/datacube-ows/blob/6d839c3a0d9a9b798b34d6ae9c5f986c75dabe5d/datacube_ows/sql/extent_views/create/011_create_time_view.sql#L13-L16

https://github.com/opendatacube/datacube-core/blob/8152f8c6682ed4d879d77735d6b2c3a755e19290/datacube/index/default-metadata-types.yaml#L193-L202

**Comment by JackLidge** [comment]
Looking at what you've posted here and also slack discussions had previously - I think the culprit as to why the yamls read correctly inside the datacube and failed to be read into the OWS was due to the reason you mentioned above: we had a `center_dt` variable in the right place but `from/to_dt` variables in the wrong place.

Why we started doing it this way I'm not sure as it was so long ago. I imagine this is probably shown better in the current documentation, which is considerably more in depth than we were working it out in late 2018.
